agent/lib/kubernetes.js

```diff
@@ -238,7 +239,22 @@ Watcher.prototype.watch = function () {
             }
         });
     });
-    request.on('error', function(e) { log.error('error on watch: %s', e); });
+    request.on('error', function(e) {
+        log.error('error on watch: %s', e);
+        if (!self.closed) {
+            request.abort();
+            setTimeout(self.list.bind(self), self.delay * 1000);
+            self.delay = Math.min(30, self.delay + 1);
+        }
+    });
+    if self.options.resyncInterval !== undefined {
+        setTimeout(function () {
```

Comment:
I think setting a timeout on the request itself (i.e. request.setTimeout(self.options.resyncInterval * 1000, function () {self.abort(); }); ) might be cleaner. The watch is generally ended by the server anyway (and would be ended by any network failure), but adding the client timeout would ensure the client set a maximum idle time, which I assume is what is wanted here.

Reply:
I was thinking to follow the same pattern of periodic re-watch as done in kubernetes controllers, as it seems more robust in cases where the expected events not to arrive (either due to a bug in the code or underlying dependencies or even bugs in the OS network layer).

Reply:
All I'm suggesting is that instead of doing your own timeout independent of the state of the request, you use the inbuilt timeout on the http agent. To me that seems more predictable and I don't as yet see any way in which it is less robust.

Simply a suggestion however.

Reply:
That makes sense, I will change it.